src/service/debtor-service.ts

```diff
@@ -0,0 +1,303 @@
+    const count = await FineHandoutEvent.count();
+
+    const records = events.map((e) => DebtorService.asBaseFineHandoutEventResponse(e));
+
+    return {
+      _pagination: {
+        take, skip, count,
+      },
+      records,
+    };
+  }
+
+  /**
+   * Return the fine handout event with the given id. Includes all its fines with the corresponding user
+   */
+  public static async getSingleFineHandoutEvent(id: number): Promise<FineHandoutEventResponse> {
+    const fineHandoutEvent = await FineHandoutEvent.findOne({
+      where: { id },
+      relations: ['fines', 'fines.userFineGroup', 'fines.userFineGroup.user'],
+      order: { createdAt: 'DESC' },
+    });
+
+    return DebtorService.asFineHandoutEventResponse(fineHandoutEvent);
+  }
+
+  /**
+   * Return all users that had at most -5 euros balance both now and on the reference date
+   * For all these users, also return their fine based on the reference date.
+   * @param userTypes List of all user types fines should be calculated for
+   * @param referenceDate Date to base fines on. If undefined, use now.
+   */
+  public static async calculateFinesOnDate({ userTypes, referenceDate }: CalculateFinesParams): Promise<UserToFineResponse[]> {
+    const debtorsOnReferenceDate = await BalanceService.getBalances({
+      maxBalance: DineroTransformer.Instance.from(-500),
+      date: referenceDate,
+      userTypes,
+    });
+
+    const debtorsNow = await BalanceService.getBalances({
+      maxBalance: DineroTransformer.Instance.from(-500),
+      userTypes,
+    });
+    const debtorsNowIds = debtorsNow.records.map((b) => b.id);
+
+    const userBalancesToFine = debtorsOnReferenceDate.records.filter((b) => debtorsNowIds.includes(b.id));
+
+    return userBalancesToFine.map((u) => {
+      const fine = calculateFine(u.amount);
+      return {
+        id: u.id,
+        amount: {
+          amount: fine.getAmount(),
+          currency: fine.getCurrency(),
+          precision: fine.getPrecision(),
+        },
+      };
+    });
+  }
+
+  /**
+   * Write fines to database for all given user ids.
+   * @param referenceDate Date to base fines on. If undefined, the date of the previous fines will be used. If this is the first fine, use now.
+   * @param userIds Ids of all users to fine
+   * @param createdBy User handing out fines
+   */
+  public static async handOutFines({
+    referenceDate, userIds,
+  }: HandOutFinesParams, createdBy: User): Promise<FineHandoutEventResponse> {
+    const previousFineGroup = (await FineHandoutEvent.find({
+      order: { id: 'desc' },
+      relations: ['fines', 'fines.userFineGroup'],
+      take: 1,
+    }))[0];
+
+    const date = referenceDate || previousFineGroup?.createdAt || new Date();
+
+    const balances = await BalanceService.getBalances({
+      date,
+      ids: userIds,
+    });
+
+    const { fines: fines1, fineHandoutEvent: fineHandoutEvent1, emails: emails1 } = await getConnection().transaction(async (manager) => {
```

Comment:
mis ergens duidelijk aangeven dat je een transaction maakt, want anders kan je vage transaction in transaction errors krijgen als iemand deze functie ook vanuit een transaction called.

Reply:
Resolved in cfcbd191b85b3707a4b3e2321c59065dc4a95ea6